drivers/ivshm/ivshm_guest_shm.c

```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: GPL-2.0
+
+#include <linux/device.h>
+#include <linux/module.h>
+#include <linux/pci.h>
+#include <linux/interrupt.h>
+#include <linux/guest_shm.h>
+
+#include "ivshm.h"
+
+#define DRIVER_VERSION	"0.01.0"
+#define DRIVER_AUTHOR	"Junjie Mao <junjie.mao@intel.com>"
+#define DRIVER_DESC	"Inter-VM shared memory driver for QEMU ivshmem devices"
+
+struct ivshm_guest_shm_dev {
+	struct ivshm_info  info;
+	struct pci_dev    *pdev;
+
+	struct guest_shm_factory __iomem *fact;
+	struct guest_shm_control *ctrl;
+
+	struct ivshm_region *iregion;
+	int                  irq;
+};
+
+static irqreturn_t guest_shm_irq_handler(int irq, void *arg)
+{
+	struct ivshm_guest_shm_dev *idev = (struct ivshm_guest_shm_dev *)arg;
+	irqreturn_t ret = IRQ_NONE;
+	int i;
+
+	if (idev->irq == irq) {
+		/* Check the status ? Only when status == (1 << id) needs handle */
+		if (idev->ctrl->status & GUEST_INTR_STATUS_MASK)
+			ivshm_notify_listeners(idev->iregion, i);
+		ret = IRQ_HANDLED;
+	}
+
+	return ret;
+}
+
+static int guest_shm_init_irq(struct ivshm_guest_shm_dev *idev)
+{
+	idev->irq = idev->fact->vector;
+	return request_irq(idev->irq, guest_shm_irq_handler, 0, "guest_shm", idev);
+}
+
+static int probe(struct pci_dev *pdev, const struct pci_device_id *id)
+{
+	struct ivshm_guest_shm_dev *idev;
+	struct ivshm_region *iregion;
+	resource_size_t paddr;
+	int err;
+	unsigned long min_align;
+	long idx;
+
+	err = pcim_enable_device(pdev);
+	if (err) {
+		dev_err(&pdev->dev, "Failed to enable PCI device.\n");
+		return err;
+	}
+
+	idev = devm_kzalloc(&pdev->dev, sizeof(struct ivshm_guest_shm_dev), GFP_KERNEL);
+	if (!idev) {
+		err = -ENOMEM;
+		goto out_disable_device;
+	}
+
+	err = pcim_iomap_regions(pdev, BIT(0), "ivshm_guest_shm");
+	if (err)
+		return err;
```

Comment:
Why we are not undoing the stuff in case of error? like pci_disable_device. You might want to look at other places in this function for similar stuffs.

Reply:
Go to 'out_disable_device' to handle it coherently.